### PR TITLE
Clipping loss

### PR DESCRIPTION
**Pull request by paulinesert**
Hi, 

I have added : 
* take into account the advantage 
* compute the probability ratios (prob_ratio function in vanilla PPO)
* compute the normalization of the advantage ( normalize_advantage in vanillaPPO)
* compute the clipping loss (can swtich from clipped to regular, in vanillaPPO)

to test : 

```Python
import jax
import chex
import haiku as hk 
from acme import specs
import jax.numpy as jnp
import numpy as np 
from jax import tree_util
import rlax
import dm_env


import dm_helper as helpers
from networks import PolicyNetwork, ValueNetwork
from agents import VanillaPPO
from replay_buffers import DataLoader

environment = helpers.make_environment("Pendulum-v1")
environment_spec = specs.make_environment_spec(environment)

FAKE_VALUE = 1.
FAKE_LOG_PROB = 0.4
batch_size = 3


from ppo.agents.base_agent import BaseAgent
from ppo.replay_buffers import FixedReplayBuffer


def policy_network(observations: chex.Array):
    return PolicyNetwork((10, 20), environment_spec.actions, "policy")(observations)
def value_network(observations: chex.Array):
    return ValueNetwork((10, 20), "value")(observations)
key = jax.random.PRNGKey(1)

agent = VanillaPPO(
        environment_spec.observations,
        policy_network,
        value_network,
        key,
        learning_rate=1e-3,
        discount=0.99,
    )

timestep = environment.reset()
agent.observe_first(timestep)
episode_return = 0
steps = 0

while not timestep.last() and steps < 9:
    action = agent.select_action(timestep.observation)
    timestep = environment.step(action)
    advantage = np.random.randn() # compute random advantage 
    agent.observe(FAKE_VALUE + steps, FAKE_LOG_PROB, action, timestep, advantage)
        
    steps += 1
    episode_return += timestep.reward

dataloader = DataLoader(agent.replay_buffer, batch_size)
dataloader.shuffle()

for transition in dataloader:
    print(transition.value_t)
    print()


policy_key, value_key = jax.random.split(key, 2)
agent.policy_network = hk.without_apply_rng(hk.transform(policy_network))
agent.policy_params = agent.policy_network.init(rng=policy_key, observations=jnp.zeros(environment_spec.observations.shape))


print("Compute ratio:", agent.prob_ratio(agent.policy_params, transition.observation_t, transition.action_t, transition.log_probability_t))
print("Compute loss clipped:", agent.policy_loss(agent.policy_params, transition.observation_t, transition.action_t, transition.advantage_t, transition.log_probability_t, 'clipped'))
print("Compute loss regular:", agent.policy_loss(agent.policy_params, transition.observation_t, transition.action_t, transition.advantage_t, 'regular'))
```

expected output : 

```
[7. 5. 9.]

[2. 1. 8.]

[3. 4. 6.]

Compute ratio: [2.3996904 1.6959783 1.362582 ]
Compute loss clipped: 0.5349039
Compute loss regular: 0.4241833
```

### CODE REVIEW

File: ppo/agents/vanilla_ppo.py [hunk 1]
```diff
@@ -29,8 +34,48 @@ def __init__(
...
+        observation,
+        action: np.ndarray,
+        advantage: chex.Array,
+        action_log_probs_old: chex.Array,
+        type: str = 'regular'
+    ) -> chex.Array:
+        mu, sigma = self.policy_network.apply(policy_params, observation)
+        action_log_probs = rlax.gaussian_diagonal().logprob(action, mu, sigma)
+
+        if type == 'clipped':
```
Comment (theovincent) [inline ppo/agents/vanilla_ppo.py:73]: Have you tried to differentiate it? I am afraid you cannot differentiate a function with an "if" loop inside.

**Comment by theovincent** [comment]
@paulinesert I think it is nice ! We can discuss about it on Wednesday. You can have a look on the branch [update](https://github.com/emasquil/ppo/tree/update), it is what we have done with @emasquil  this afternoon :)
